pkg/controllers/namespace/namespaceController.go

```diff
@@ -0,0 +1,268 @@
+					glog.Errorf("Error while converting %#v to calico profile.", newObj)
+					return
+				}
+				// Add profileName:profile in calicoCache
+				ccache.Set(profile.(*api.Profile).Metadata.Name, profile)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+			// key function.
+
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			glog.Infof("Got DELETE event for namespace: %s\n", key)
+
+			if err == nil {
+				calicoKey := fmt.Sprintf(converter.ProfileNameFormat+"%s", key)
+				ccache.Delete(calicoKey)
+			}
+		},
+	}, cache.Indexers{})
+
+	ccache.SetIndexer(indexer)
+
+	return &NamespaceController{indexer, informer, ccache, cclient}
+}
+
+// Run starts controller.Internally it starts syncing
+// kubernetes and calico caches.
+func (c *NamespaceController) Run(threadiness int, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	// Let the workers stop when we are done
+	workqueue := c.calicoObjCache.GetQueue()
+	defer workqueue.ShutDown()
+
+	glog.Info("Starting namespace controller")
+
+	// Start Calico cache. Cache gets loaded with objects
+	// from ETCD datastore.
+	c.calicoObjCache.Run(stopCh)
+
+	go c.informer.Run(stopCh)
+
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		uruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	// Start a number of worker threads to read from the queue.
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("Stopping Node controller")
+}
+
+func (c *NamespaceController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+func (c *NamespaceController) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	workqueue := c.calicoObjCache.GetQueue()
+	key, quit := workqueue.Get()
+	if quit {
+		return false
+	}
+	// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+	// This allows safe parallel processing because two nodes with the same key are never processed in
+	// parallel.
+	defer workqueue.Done(key)
+
+	// Invoke the method containing the business logic
+	err := c.syncToCalico(key.(string))
+
+	// Handle the error if something went wrong during the execution of the business logic
+	c.handleErr(err, key.(string))
+	return true
+}
+
+// syncToCalico syncs the given update to Calico's etcd, as well as the in-memory cache
+// of Calico objects.
+func (c *NamespaceController) syncToCalico(key string) error {
+
+	// Check if it exists in our cache.
+	obj, exists := c.calicoObjCache.Get(key)
+
+	if !exists {
+
+		glog.Debugf("namespace %s does not exist anymore on kubernetes\n", key)
+		glog.Infof("Deleting namespace %s on ETCD \n", key)
+
+		err := c.calicoClient.Profiles().Delete(api.ProfileMetadata{
```

Comment:
I think delete needs to be idempotent.

i.e. if it doesn't exist in etcd, we treat it as if we successfully deleted it.